sklearn/feature_selection/_base.py

```diff
@@ -121,3 +122,42 @@ def inverse_transform(self, X):
         Xt = np.zeros((X.shape[0], support.size), dtype=X.dtype)
         Xt[:, support] = X
         return Xt
+
+
+def _get_importances_auto(estimator):
```

Comment:
we can integrate this function within a `_get_raw_feature_importances` and have `_get_feature_importances` calling it to take the norm or the abs. In the other function, RFE class we only have to call `_get_raw_feature_importances` which return directly the importance and make the validation if the this `auto` or not.

Reply:
Sorry, This suggestion is not quite clear to me. To you mean to say `_get_importances_auto` when you are mentioning `_get_raw_feature_importances`?

Reply:
When we put these two functions together, we might have to find a better way to handle to error message handling as well.

Reply:
I mean a something like this:

```diff
diff --git a/sklearn/feature_selection/_base.py b/sklearn/feature_selection/_base.py
index 67c791dd7..faa9a4b6f 100644
--- a/sklearn/feature_selection/_base.py
+++ b/sklearn/feature_selection/_base.py
@@ -124,35 +124,33 @@ class SelectorMixin(TransformerMixin, metaclass=ABCMeta):
         return Xt
 
 
-def _get_importances_auto(estimator):
-    if hasattr(estimator, 'coef_'):
-        getter = attrgetter('coef_')
-    elif hasattr(estimator, 'feature_importances_'):
-        getter = attrgetter('feature_importances_')
-    else:
-        raise ValueError(
-            f"when `importance_getter=='auto'`, the underlying estimator "
-            f"{estimator.__class__.__name__} should have `coef_` or "
-            f"`feature_importances_` attribute. Either pass a fitted "
-            f"estimator to feature selector or call fit before calling "
-            f"transform."
-        )
-
-    return getter
-
-
-def _get_feature_importances(estimator, getter,
-                             norm_order=1):
+def _get_raw_feature_importances(estimator, getter):
     """Retrieve or aggregate feature importances from estimator"""
     if isinstance(getter, str):
         if getter == 'auto':
-            getter = _get_importances_auto(estimator)
+            if hasattr(estimator, 'coef_'):
+                getter = attrgetter('coef_')
+            elif hasattr(estimator, 'feature_importances_'):
+                getter = attrgetter('feature_importances_')
+            else:
+                raise ValueError(
+                    f"when `importance_getter=='auto'`, the underlying "
+                    f"estimator {estimator.__class__.__name__} should have "
+                    f"`coef_` or `feature_importances_` attribute. Either "
+                    f"pass a fitted estimator to feature selector or call fit "
+                    f"before calling transform."
+                )
         else:
             getter = attrgetter(getter)
     elif not callable(getter):
-        raise ValueError('`importance_getter` has to be a string'
-                         ' or `callable`')
-    importances = getter(estimator)
+        raise ValueError(
+            '`importance_getter` has to be a string or `callable`'
+        )
+    return getter(estimator)
+
+
+def _get_feature_importances(estimator, getter, norm_order=1):
+    importances = _get_raw_feature_importances(estimator, getter)
 
     if importances.ndim == 1:
         importances = np.abs(importances)
diff --git a/sklearn/feature_selection/_rfe.py b/sklearn/feature_selection/_rfe.py
index 4e2b5962d..e5ab23e56 100644
--- a/sklearn/feature_selection/_rfe.py
+++ b/sklearn/feature_selection/_rfe.py
@@ -8,7 +8,6 @@
 
 import numpy as np
 from joblib import Parallel, delayed, effective_n_jobs
-from operator import attrgetter
 
 from ..utils import safe_sqr
 from ..utils.metaestimators import if_delegate_has_method
@@ -22,7 +21,8 @@ from ..base import is_classifier
 from ..model_selection import check_cv
 from ..model_selection._validation import _score
 from ..metrics import check_scoring
-from ._base import SelectorMixin, _get_importances_auto
+from ._base import SelectorMixin
+from ._base import _get_raw_feature_importances
 
 
 def _rfe_single_fit(rfe, estimator, X, y, train, test, scorer):
@@ -213,16 +213,9 @@ class RFE(SelectorMixin, MetaEstimatorMixin, BaseEstimator):
             estimator.fit(X[:, features], y)
 
             # Get feature importance
-            importance_getter = self.importance_getter
-            if isinstance(importance_getter, str):
-                if importance_getter == 'auto':
-                    importance_getter = _get_importances_auto(estimator)
-                else:
-                    importance_getter = attrgetter(importance_getter)
-            elif not callable(importance_getter):
-                raise ValueError('`importance_getter` has to be a string'
-                                 ' or `callable`')
-            importances = importance_getter(estimator)
+            importances = _get_raw_feature_importances(
+                estimator, self.importance_getter
+            )
 
             # Get ranks
             if importances.ndim > 1:
```

Reply:
Basically it should not change anything about the error handling, you should be fine with that.

Reply:
even better is to code the rank in the `_get_feature_importances indeed`

Reply:
This would give something like this:

``` diff
diff --git a/sklearn/feature_selection/_base.py b/sklearn/feature_selection/_base.py
index 67c791dd7..982ce981b 100644
--- a/sklearn/feature_selection/_base.py
+++ b/sklearn/feature_selection/_base.py
@@ -12,7 +12,9 @@ import numpy as np
 from scipy.sparse import issparse, csc_matrix
 
 from ..base import TransformerMixin
-from ..utils import check_array, safe_mask
+from ..utils import check_array
+from ..utils import safe_mask
+from ..utils import safe_sqr
 
 
 class SelectorMixin(TransformerMixin, metaclass=ABCMeta):
@@ -124,40 +126,43 @@ class SelectorMixin(TransformerMixin, metaclass=ABCMeta):
         return Xt
 
 
-def _get_importances_auto(estimator):
-    if hasattr(estimator, 'coef_'):
-        getter = attrgetter('coef_')
-    elif hasattr(estimator, 'feature_importances_'):
-        getter = attrgetter('feature_importances_')
-    else:
-        raise ValueError(
-            f"when `importance_getter=='auto'`, the underlying estimator "
-            f"{estimator.__class__.__name__} should have `coef_` or "
-            f"`feature_importances_` attribute. Either pass a fitted "
-            f"estimator to feature selector or call fit before calling "
-            f"transform."
-        )
-
-    return getter
-
-
-def _get_feature_importances(estimator, getter,
+def _get_feature_importances(estimator, getter, ranking_func="square",
                              norm_order=1):
     """Retrieve or aggregate feature importances from estimator"""
     if isinstance(getter, str):
         if getter == 'auto':
-            getter = _get_importances_auto(estimator)
+            if hasattr(estimator, 'coef_'):
+                getter = attrgetter('coef_')
+            elif hasattr(estimator, 'feature_importances_'):
+                getter = attrgetter('feature_importances_')
+            else:
+                raise ValueError(
+                    f"when `importance_getter=='auto'`, the underlying "
+                    f"estimator {estimator.__class__.__name__} should have "
+                    f"`coef_` or `feature_importances_` attribute. Either "
+                    f"pass a fitted estimator to feature selector or call fit "
+                    f"before calling transform."
+                )
         else:
             getter = attrgetter(getter)
     elif not callable(getter):
-        raise ValueError('`importance_getter` has to be a string'
-                         ' or `callable`')
+        raise ValueError(
+            '`importance_getter` has to be a string or `callable`'
+        )
     importances = getter(estimator)
 
-    if importances.ndim == 1:
-        importances = np.abs(importances)
+    if ranking_func == "norm":
+        if importances.ndim == 1:
+            importances = np.abs(importances)
+        else:
+            importances = np.linalg.norm(importances, axis=0,
+                                         ord=norm_order)
+    elif ranking_func == "square":
+        if importances.ndim > 1:
+            ranks = np.argsort(safe_sqr(importances).sum(axis=0))
+        else:
+            ranks = np.argsort(safe_sqr(importances))
     else:
-        importances = np.linalg.norm(importances, axis=0,
-                                     ord=norm_order)
+        raise ValueError("XXXX")
 
     return importances
diff --git a/sklearn/feature_selection/_rfe.py b/sklearn/feature_selection/_rfe.py
index 4e2b5962d..e5ab23e56 100644
--- a/sklearn/feature_selection/_rfe.py
+++ b/sklearn/feature_selection/_rfe.py
@@ -8,7 +8,6 @@
 
 import numpy as np
 from joblib import Parallel, delayed, effective_n_jobs
-from operator import attrgetter
 
 from ..utils import safe_sqr
 from ..utils.metaestimators import if_delegate_has_method
@@ -22,7 +21,8 @@ from ..base import is_classifier
 from ..model_selection import check_cv
 from ..model_selection._validation import _score
 from ..metrics import check_scoring
-from ._base import SelectorMixin, _get_importances_auto
+from ._base import SelectorMixin
+from ._base import _get_raw_feature_importances
 
 
 def _rfe_single_fit(rfe, estimator, X, y, train, test, scorer):
@@ -213,16 +213,9 @@ class RFE(SelectorMixin, MetaEstimatorMixin, BaseEstimator):
             estimator.fit(X[:, features], y)
 
             # Get feature importance
-            importance_getter = self.importance_getter
-            if isinstance(importance_getter, str):
-                if importance_getter == 'auto':
-                    importance_getter = _get_importances_auto(estimator)
-                else:
-                    importance_getter = attrgetter(importance_getter)
-            elif not callable(importance_getter):
-                raise ValueError('`importance_getter` has to be a string'
-                                 ' or `callable`')
-            importances = importance_getter(estimator)
+            importances = _get_raw_feature_importances(
+                estimator, self.importance_getter
+            )
 
             # Get ranks
             if importances.ndim > 1:
```

Reply:
This is a pitty that the criterion for `SelectFromModel` and the `RFE` are different. I don't know if it would make sense to have something homogeneous. I would say that this is not the goal for this PR anyway.

Reply:
Do these changes seem OK with you?

Reply:
Thanks a lot for your inputs. The suggestion is clear to me now. Only concern I have is that we are applying the transformations on the feature importance but we are naming the function as `get_feature_importances`. can we rename it ?

Reply:
We could rename it since this is a private function. We already were applying a transform in the passed. However, from your comment, we could rename `ranking_func` with another name like `transform_func`. We could add the option `transform_func=None` by default that return the `feature_importances` without transformation.

This would be more explicit indeed.